internal/gatewayapi/route.go

```diff
@@ -493,30 +493,31 @@ func (t *Translator) processGRPCRouteRules(grpcRoute *GRPCRouteContext, parentRe
 
 		for _, backendRef := range rule.BackendRefs {
 			backendRef := backendRef
-			ds, backendWeight := t.processDestination(backendRef, parentRef, grpcRoute, resources)
+			ds := t.processDestination(backendRef, parentRef, grpcRoute, resources)
+			if ds == nil {
+				continue
+			}
+
 			for _, route := range ruleRoutes {
 				// If the route already has a direct response or redirect configured, then it was from a filter so skip
 				// processing any destinations for this route.
-				if route.DirectResponse == nil && route.Redirect == nil {
-					if ds != nil && len(ds.Endpoints) > 0 {
-						if route.Destination == nil {
-							route.Destination = &ir.RouteDestination{
-								Name: irRouteDestinationName(grpcRoute, ruleIdx),
-							}
-						}
-						route.Destination.Settings = append(route.Destination.Settings, ds)
-						route.BackendWeights.Valid += backendWeight
+				if route.DirectResponse != nil || route.Redirect != nil {
+					continue
+				}
 
-					} else {
-						route.BackendWeights.Invalid += backendWeight
+				if route.Destination == nil {
+					route.Destination = &ir.RouteDestination{
+						Name: irRouteDestinationName(grpcRoute, ruleIdx),
 					}
 				}
+				route.Destination.Settings = append(route.Destination.Settings, ds)
 			}
 		}
 
 		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
 		for _, ruleRoute := range ruleRoutes {
-			if ruleRoute.Destination == nil && ruleRoute.Redirect == nil {
+			noValidBackends := ruleRoute.Destination == nil || ruleRoute.Destination.ToBackendWeights().Valid == 0
+			if noValidBackends && ruleRoute.Redirect == nil {
```

Comment:
isnt something like this needed in `processHTTPRouteRules` as well ?

Reply:
good catch